File: benchmark/zig_benchmark/src/debug_tree_building_step_by_step.zig
const std = @import("std");
const log = @import("hash-zig").utils.log;
const hash_zig = @import("hash-zig");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const seed_hex = std.process.getEnvVarOwned(allocator, "SEED_HEX") catch "4242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242";
    defer allocator.free(seed_hex);

    log.print("=== Zig Tree Building Step-by-Step Analysis ===\n", .{});
    log.print("SEED: {s}\n", .{seed_hex});

    // Parse seed
    const seed_bytes = try std.fmt.allocPrint(allocator, "{s}", .{seed_hex});
    defer allocator.free(seed_bytes);

    var seed_array: [32]u8 = undefined;
    for (0..32) |i| {
        const hex_pair = seed_bytes[i * 2 .. i * 2 + 2];
        seed_array[i] = std.fmt.parseInt(u8, hex_pair, 16) catch unreachable;
    }

    var scheme = try hash_zig.GeneralizedXMSSSignatureScheme.initWithSeed(allocator, .lifetime_2_8, seed_array);
    defer scheme.deinit();

    log.print("\n=== Key Generation ===\n", .{});
    const result = try scheme.keyGen(0, 256);
    defer result.secret_key.deinit();

    log.print("Final root values: {any}\n", .{result.public_key.root});
    log.print("Parameter values: {any}\n", .{result.public_key.parameter});

    // Show the exact values in hex
    log.print("\n=== Hex Values ===\n", .{});
    log.print("Root values (hex):\n", .{});
    for (result.public_key.root, 0..) |val, i| {
        log.print("  [{}] = 0x{x} ({})\n", .{ i, val.value, val.value });
    }

    log.print("Parameter values (hex):\n", .{});
    for (result.public_key.parameter, 0..) |val, i| {
        log.print("  [{}] = 0x{x} ({})\n", .{ i, val.value, val.value });
    }
}

File: benchmark/rust_benchmark/src/bin/debug_tree_building_step_by_step.rs
use leansig::signature::generalized_xmss::instantiations_poseidon_top_level::lifetime_2_to_the_8::SIGTopLevelTargetSumLifetime8Dim64Base8;
use leansig::signature::SignatureScheme;
use rand::rngs::StdRng;
use rand::SeedableRng;
use std::env;

fn main() {
    let seed_hex = env::var("SEED_HEX").unwrap_or_else(|_| "4242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242424242".to_string());

    println!("=== Rust Tree Building Step-by-Step Debug ===");
    println!("SEED: {}", seed_hex);

    // Parse seed
    let seed_bytes = hex::decode(&seed_hex).expect("Invalid hex seed");
    let mut seed_array = [0u8; 32];
    seed_array.copy_from_slice(&seed_bytes[..32]);

    let mut rng = StdRng::from_seed(seed_array);

    println!("\n=== Generate Key and Extract Details ===");

    // Generate a key to get the exact parameters and root
    let (pk, _sk) = SIGTopLevelTargetSumLifetime8Dim64Base8::key_gen(&mut rng, 0, 256);
    let pk_json = serde_json::to_string(&pk).unwrap();

    // Parse the JSON to extract values
    let pk_data: serde_json::Value = serde_json::from_str(&pk_json).unwrap();
    let param_array = pk_data["parameter"].as_array().unwrap();
    let root_array = pk_data["root"].as_array().unwrap();

    let mut param_values = [0u32; 5];
    for (i, val) in param_array.iter().enumerate() {
        param_values[i] = val.as_u64().unwrap() as u32;
    }

    let mut root_values = [0u32; 8];
    for (i, val) in root_array.iter().enumerate() {
        root_values[i] = val.as_u64().unwrap() as u32;
    }

    println!("Parameter values: {:?}", param_values);
    println!("Root values: {:?}", root_values);

    // Test the specific tree hash operation that should produce 0x31461cb0
    println!("\n=== Test Specific Tree Hash Operation ===");

    // These are the exact values from the Zig debug output
    let left_child = 0x1640cb16;
    let right_child = 0x54503ce2;
    let level: u8 = 5;
    let pos_in_level: u32 = 0;

    println!("Left child: 0x{:x} ({})", left_child, left_child);
    println!("Right child: 0x{:x} ({})", right_child, right_child);
    println!("Level: {}, Position: {}", level, pos_in_level);

    // Test tweak computation
    let tweak_bigint = ((level as u128) << 40) | ((pos_in_level as u128) << 8) | 0x01;
    let p: u128 = 2130706433; // KoalaBear field modulus
    let tweak = [(tweak_bigint % p) as u32, ((tweak_bigint / p) % p) as u32];

    println!("Tweak bigint: 0x{:x}", tweak_bigint);
    println!("Tweak[0]: 0x{:x} ({})", tweak[0], tweak[0]);
    println!("Tweak[1]: 0x{:x} ({})", tweak[1], tweak[1]);

    // Show the complete input that will be passed to Poseidon2
    println!("\nComplete Poseidon2 input (parameter + tweak + message):");
    println!("Parameter (5 elements):");
    for (i, val) in param_values.iter().enumerate() {
        println!("  [{}] = 0x{:x}", i, val);
    }
    println!("Tweak (2 elements):");
    for (i, val) in tweak.iter().enumerate() {
        println!("  [{}] = 0x{:x}", i, val);
    }
    println!("Message (16 elements):");
    println!("  [0] = 0x{:x}", left_child);
    println!("  [1] = 0x{:x}", right_child);
    println!("  [2-15] = 0x0 (padding)");

    println!("\nExpected result from Zig debug: 0x31461cb0");
    println!("Actual Rust root[0]: 0x{:x}", root_values[0]);
    println!("Match: {}", root_values[0] == 0x31461cb0);

    // Test if the issue is in the tree building algorithm by comparing intermediate results
    println!("\n=== Tree Building Algorithm Analysis ===");
    println!("The issue appears to be in the tree building algorithm itself.");
    println!("First hash operation should produce 0x31461cb0, but subsequent operations diverge.");
    println!("This suggests the tree building logic differs between Rust and Zig implementations.");
}
